src/lobpcg.jl

```diff
@@ -0,0 +1,751 @@
+XBR!(BlockGram, XBlocks, RBlocks, n) = Ac_mul_B!(view(BlockGram.XAR, :, 1:n), XBlocks.block, view(RBlocks.B_block, :, 1:n))
+RBP!(BlockGram, RBlocks, PBlocks, n) = Ac_mul_B!(view(BlockGram.RAP, 1:n, 1:n), view(RBlocks.B_block, :, 1:n), view(PBlocks.block, :, 1:n))
+XBX!(BlockGram, XBlocks) = Ac_mul_B!(BlockGram.XAX, XBlocks.block, XBlocks.B_block)
+RBR!(BlockGram, RBlocks, n) = Ac_mul_B!(view(BlockGram.RAR, 1:n, 1:n), view(RBlocks.block, :, 1:n), view(RBlocks.B_block, :, 1:n))
+PBP!(BlockGram, PBlocks, n) = Ac_mul_B!(view(BlockGram.PAP, 1:n, 1:n), view(PBlocks.block, :, 1:n), view(PBlocks.B_block, :, 1:n))
+
+function I!(G, xr)
+    for j in xr, i in xr
+        G[i, j] = ifelse(i==j, 1, 0)
+    end
+    return
+end
+
+function (g::BlockGram)(gram, lambda, n1::Int, n2::Int, n3::Int)
+    xr = 1:n1
+    rr = n1+1:n1+n2
+    pr = n1+n2+1:n1+n2+n3
+    if n1 > 0
+        #gram[xr, xr] .= view(g.XAX, 1:n1, 1:n1)
+        gram[xr, xr] .= Diagonal(view(lambda, 1:n1))
+    end
+    if n2 > 0
+        gram[rr, rr] .= view(g.RAR, 1:n2, 1:n2)
+        gram[xr, rr] .= view(g.XAR, 1:n1, 1:n2)
+        conj!(transpose!(view(gram, rr, xr), view(g.XAR, 1:n1, 1:n2)))
+    end
+    if n3 > 0
+        gram[pr, pr] .= view(g.PAP, 1:n3, 1:n3)
+        gram[rr, pr] .= view(g.RAP, 1:n2, 1:n3)
+        gram[xr, pr] .= view(g.XAP, 1:n1, 1:n3)
+        conj!(transpose!(view(gram, pr, rr), view(g.RAP, 1:n2, 1:n3)))
+        conj!(transpose!(view(gram, pr, xr), view(g.XAP, 1:n1, 1:n3)))
+    end
+    return 
+end
+function (g::BlockGram)(gram, n1::Int, n2::Int, n3::Int, normalized::Bool=true)
+    xr = 1:n1
+    rr = n1+1:n1+n2
+    pr = n1+n2+1:n1+n2+n3
+    if n1 > 0
+        if normalized
+            I!(gram, xr)
+        else
+            gram[xr, xr] .= view(g.XAX, 1:n1, 1:n1)
+        end
+    end
+    if n2 > 0
+        if normalized
+            I!(gram, rr)
+        else
+            gram[rr, rr] .= view(g.RAR, 1:n2, 1:n2)
+        end
+        gram[xr, rr] .= view(g.XAR, 1:n1, 1:n2)
+        conj!(transpose!(view(gram, rr, xr), view(g.XAR, 1:n1, 1:n2)))
+    end
+    if n3 > 0
+        if normalized
+            I!(gram, pr)
+        else
+            gram[pr, pr] .= view(g.PAP, 1:n3, 1:n3)
+        end
+        gram[rr, pr] .= view(g.RAP, 1:n2, 1:n3)
+        gram[xr, pr] .= view(g.XAP, 1:n1, 1:n3)
+        conj!(transpose!(view(gram, pr, rr), view(g.RAP, 1:n2, 1:n3)))
+        conj!(transpose!(view(gram, pr, xr), view(g.XAP, 1:n1, 1:n3)))
+    end
+    return 
+end
+
+abstract type AbstractOrtho end
+struct CholQR{TA} <: AbstractOrtho
+    gramVBV::TA # to be used in view
+end
+
+function A_rdiv_B!(A, B::UpperTriangular)
+    s = size(A, 2)
+    A[:,1] .= view(A, :, 1) ./ B[1,1]
+    for i in 2:s
+        for j in 1:i-1
+            A[:,i] .= view(A, :, i) - view(A, :, j) .* B[j,i]
+        end
+        A[:,i] .= view(A, :, i) ./ B[i,i]
+    end
+    return
```

Comment:
Convention is to return `A`.

Reply:
`@inbounds`? This is an O(n) operation right? Better make it fast -- although there's a blas wrapper for this on 0.7.

Reply:
Sure.